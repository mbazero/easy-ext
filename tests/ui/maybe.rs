// See also `maybe` test in test.rs

use easy_ext::ext;

#[ext(E1)]
impl<T: ?Sized> T // Ok
{
    fn f(&self) {}
}

#[ext(E2)]
impl<T> T
where
    T: ?Sized, // Ok
{
    fn f(&self) {}
}

#[ext(E3)]
impl<T> T
where
    Self: ?Sized, //~ ERROR `?Trait` bounds are only permitted at the point where a type parameter is declared
{
    fn f(&self) {}
}

// The following is a case where #[ext] is not used. The behavior should match in both cases.

trait T1
where
    Self: ?Sized,
{
    //~^^ ERROR `?Trait` bounds are only permitted at the point where a type parameter is declared
    fn f(&self);
}

trait T2 {
    fn f(&self);
}
impl<T> T2 for T
where
    Self: ?Sized, //~ ERROR `?Trait` bounds are only permitted at the point where a type parameter is declared
{
    fn f(&self) {}
}

trait T3 {
    fn f(&self);
}
impl<T: ?Sized> T3 for T // Ok
{
    fn f(&self) {}
}

trait T4 {
    fn f(&self);
}
impl<T> T4 for T
where
    T: ?Sized, // Ok
{
    fn f(&self) {}
}

trait T5 {
    fn f(&self);
}
impl<T> T5 for T {
    fn f(&self)
    where
        T: ?Sized, //~ ERROR `?Trait` bounds are only permitted at the point where a type parameter is declared
    {
    }
}

fn main() {}
