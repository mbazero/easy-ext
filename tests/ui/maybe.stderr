error: `?Trait` bounds are only permitted at the point where a type parameter is declared
  --> $DIR/maybe.rs:22:5
   |
22 |     Self: ?Sized, //~ ERROR `?Trait` bounds are only permitted at the point where a type parameter is declared
   |     ^^^^

error: `?Trait` bounds are only permitted at the point where a type parameter is declared
  --> $DIR/maybe.rs:31:5
   |
31 |     Self: ?Sized,
   |     ^^^^

error: `?Trait` bounds are only permitted at the point where a type parameter is declared
  --> $DIR/maybe.rs:42:5
   |
42 |     Self: ?Sized, //~ ERROR `?Trait` bounds are only permitted at the point where a type parameter is declared
   |     ^^^^

error: `?Trait` bounds are only permitted at the point where a type parameter is declared
  --> $DIR/maybe.rs:71:9
   |
71 |         T: ?Sized, //~ ERROR `?Trait` bounds are only permitted at the point where a type parameter is declared
   |         ^
